syntax = "proto3";
import "google/protobuf/struct.proto";

package docarray;

/**
 * Represents a (quantized) dense n-dim array
 */
message DenseNdArrayProto {
  // the actual array data, in bytes
  bytes buffer = 1;

  // the shape (dimensions) of the array
  repeated uint32 shape = 2;

  // the data type of the array
  string dtype = 3;
}

/**
 * Represents a general n-dim array, can be either dense or sparse
 */
message NdArrayProto {
  DenseNdArrayProto dense = 1; // dense representation of the ndarray

  google.protobuf.Struct parameters = 2;
}


//
message NodeProto {

  oneof content {
    bytes blob = 1;
    // the ndarray of the image/audio/video document
    NdArrayProto ndarray = 2;
    // a text
    string text = 3;
    // a sub Document
    DocumentProto document = 4;
    // a sub DocumentArray
    DocumentArrayProto document_array = 5;
  }

  string docarray_type = 6;

}

/**
 * Represents a Document
 */
message DocumentProto {

  map<string, NodeProto> data = 1;

}

message DocumentArrayProto {
    repeated DocumentProto docs = 1; // a list of Documents
}

message UnionArrayProto { // represent the column of the da stacked
    oneof content {
        DocumentArrayStackedProto document_array = 1;
        NdArrayProto ndarray = 2;
    }
}

message DocumentArrayStackedProto{
  DocumentArrayProto list_ = 1; // a list of Documents
  map<string, UnionArrayProto> columns = 2; // a dict of columns
}